lib/puppet/network/http.rb

```diff
@@ -5,6 +5,10 @@ module Puppet::Network::HTTP
   MASTER_URL_PREFIX = "/puppet"
   CA_URL_PREFIX = "/puppet-ca"
 
+  NOT_FOUND_ERROR_MESSAGE = 'This master is running Puppet 4, and the URL format has changed. ' +
+      'This request was likely made with a Puppet 3 Agent, as it used an invalid ' +
```

Comment:
Should the message report PUPPETVERSION rather than a fixed string of "Puppet 4". E.g. if someone has puppet 5 this would be confusing. But I could also be persuaded that we can cross that bridge when we come to it.